hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsHttpConnection.java

```diff
@@ -0,0 +1,367 @@
+   * @param length the number of bytes to be written to the buffer.
+   *
+   * @throws IOException if an error occurs.
+   */
+  public void processResponse(byte[] buffer, final int offset,
+      final int length) throws IOException {
+    // get the response
+    long startTime = 0;
+    if (isTraceEnabled()) {
+      startTime = System.nanoTime();
+    }
+
+    setStatusCode(this.connection.getResponseCode());
+
+    if (isTraceEnabled()) {
+      setRecvResponseTimeMs(elapsedTimeMs(startTime));
+    }
+
+    setStatusDescription(this.connection.getResponseMessage());
+
+    setRequestId(this.connection.getHeaderField(
+        HttpHeaderConfigurations.X_MS_REQUEST_ID));
+    if (getRequestId() == null) {
+      setRequestId(AbfsHttpConstants.EMPTY_STRING);
+    }
+    // dump the headers
+    AbfsIoUtils.dumpHeadersToDebugLog("Response Headers",
+        connection.getHeaderFields());
+
+    if (AbfsHttpConstants.HTTP_METHOD_HEAD.equals(getMethod())) {
+      // If it is HEAD, and it is ERROR
+      return;
+    }
+
+    if (isTraceEnabled()) {
+      startTime = System.nanoTime();
+    }
+
+    if (getStatusCode() >= HttpURLConnection.HTTP_BAD_REQUEST) {
+      processStorageErrorResponse();
+      if (isTraceEnabled()) {
+        setRecvResponseTimeMs(getRecvResponseTimeMs() + elapsedTimeMs(startTime));
+      }
+      setBytesReceived(this.connection.getHeaderFieldLong(
+          HttpHeaderConfigurations.CONTENT_LENGTH, 0));
+    } else {
+      // consume the input stream to release resources
+      int totalBytesRead = 0;
+
+      try (InputStream stream = this.connection.getInputStream()) {
+        if (isNullInputStream(stream)) {
+          return;
+        }
+        boolean endOfStream = false;
+
+        // this is a list operation and need to retrieve the data
+        // need a better solution
+        if (AbfsHttpConstants.HTTP_METHOD_GET.equals(getMethod())
+            && buffer == null) {
+          parseListFilesResponse(stream);
+        } else if (AbfsHttpConstants.HTTP_METHOD_POST.equals(getMethod())) {
+          int contentLen = this.connection.getContentLength();
+          if (contentLen != 0) {
+            try (DataInputStream dis = new DataInputStream(stream)) {
+              byte[] contentBuffer  = new byte[contentLen];
+              dis.readFully(contentBuffer);
+              setResponseContentBuffer(contentBuffer);
+              totalBytesRead += contentLen;
+            }
+          }
+        } else {
+          if (buffer != null) {
+            while (totalBytesRead < length) {
+              int bytesRead = stream.read(buffer, offset + totalBytesRead,
+                  length - totalBytesRead);
+              if (bytesRead == -1) {
+                endOfStream = true;
+                break;
+              }
+              totalBytesRead += bytesRead;
+            }
+          }
+          if (!endOfStream && stream.read() != -1) {
```

Comment:
I couldnt find the cause from the commit history, however I noticed many forums that suggest issues with data still being stream even though the stream was closed at client side. Hence believe this is to avoid any issues for that issues.
Have addressed the comment to add a new metric into IOStatistic to track bytes that are being discarded in [HADOOP-17890 PR](https://github.com/apache/hadoop/pull/3381).